### PR TITLE
GS/TC: Kill 16bit target if invalidating non 16bit

### PR DESCRIPTION
**Pull request by refractionpcsx2**
### Description of Changes
Kill fundamentally incompatible render targets

### Rationale behind Changes
If the formatting is completely different, then it's highly unlikely it's going to be used, however this is kind of a fudge to fix FFX-2, which translates the Z buffer as C16 (which it does right the first time), then goes back to using it as a Z buffer (not sure this gets put back, but irrelevant for now), this means a C16 target is hanging around, this gets invalidated as Z24 is uploaded again, then it uses it as a target again, but instead of getting the newly updated Z24 format, it find the old C16 target, makes a mess of updating the target (because of the format nonsense which we don't handle in the HW renderer), and causes stripes down the screen.

Ideally we want to kill the C16 target once we're finished with it so it's not hanging around, it's from a previous frame so there's no real need to keep it.

### Suggested Testing Steps
Try Random games + FFX-2 menus.  Dump run is okay, a couple of games showed up from target size changes (presumably shrunk), but that was it.

FFX-2
Master:
image

PR:
image

### CODE REVIEW

File: pcsx2/GS/Renderers/HW/GSTextureCache.cpp [hunk 1]
```diff
@@ -389,9 +389,17 @@ void GSTextureCache::DirtyRectByPage(u32 sbp, u32 spsm, u32 sbw, Target* t, GSVe
...
+	// It might be better to get rid of targets that have been translated from Z once we're finished with them, however this was essentially the next draw, but got invalidated (badly).
+	// FIXME I guess, I don't like this solution, it shouldn't be the invalidations job to manage draws swapping around Z and RT's.
+	bool incompatible_formats = (GSLocalMemory::m_psm[spsm].bpp == 16) != (GSLocalMemory::m_psm[t->m_TEX0.PSM].bpp == 16 && !t->m_32_bits_fmt);
+	incompatible_formats = incompatible_formats && GSUtil::GetChannelMask(spsm) != GSUtil::GetChannelMask(t->m_TEX0.PSM);
+	// If the whole thing is covered, just invalidate the whole rect or it's a 16bit target trying to take a non 16bit value (fundamentally incompatible).
+	if ((start_bp <= t->m_TEX0.TBP0 && end_bp >= t->UnwrappedEndBlock()) || incompatible_formats)
+	//if ((start_bp <= t->m_TEX0.TBP0 && end_bp >= t->UnwrappedEndBlock()) || ((GSLocalMemory::m_psm[spsm].bpp == 16) != (GSLocalMemory::m_psm[t->m_TEX0.PSM].bpp == 16 && !t->m_32_bits_fmt)))
+	{
+		if(!(start_bp <= t->m_TEX0.TBP0 && end_bp >= t->UnwrappedEndBlock()))
```
Comment (lightningterror) [inline pcsx2/GS/Renderers/HW/GSTextureCache.cpp:401]: Space after if.

Comment (refractionpcsx2) [reply pcsx2/GS/Renderers/HW/GSTextureCache.cpp]: That line wasn't supposed to be there anyway.

File: pcsx2/GS/Renderers/HW/GSTextureCache.cpp [hunk 2]
```diff
@@ -389,8 +389,13 @@ void GSTextureCache::DirtyRectByPage(u32 sbp, u32 spsm, u32 sbw, Target* t, GSVe
 	const u32 end_bp = GSLocalMemory::GetEndBlockAddress(sbp, sbw, spsm, src_r);
 	GL_INS("Invalidating BP: 0x%x (%x -> %x) BW: %d PSM %s Target BP: 0x%x BW %x PSM %s", sbp, start_bp, end_bp, sbw, psm_str(spsm), t->m_TEX0.TBP0, t->m_TEX0.TBW, psm_str(t->m_TEX0.PSM));
 
-	// If the whole thing is covered, just invalidate the whole rect.
-	if (start_bp <= t->m_TEX0.TBP0 && end_bp >= t->UnwrappedEndBlock())
+	// This was added basically for FFX-2 which does some shuffling with the Z buffer, causing it to later be picked up again.
+	// It might be better to get rid of targets that have been translated from Z once we're finished with them, however this was essentially the next draw, but got invalidated (badly).
+	// FIXME I guess, I don't like this solution, it shouldn't be the invalidations job to manage draws swapping around Z and RT's.
+	bool incompatible_formats = (GSLocalMemory::m_psm[spsm].bpp == 16) != (GSLocalMemory::m_psm[t->m_TEX0.PSM].bpp == 16 && !t->m_32_bits_fmt);
+	incompatible_formats = incompatible_formats && GSUtil::GetChannelMask(spsm) != GSUtil::GetChannelMask(t->m_TEX0.PSM);
```
Comment (lightningterror) [inline pcsx2/GS/Renderers/HW/GSTextureCache.cpp:396]: Can do `incompatible_formats &= GSUtil::GetChannelMask(spsm) != GSUtil::GetChannelMask(t->m_TEX0.PSM);` instead.
Can also make it part of the previous line as well.

Comment (refractionpcsx2) [reply pcsx2/GS/Renderers/HW/GSTextureCache.cpp:396]: No you cannot.

Comment (refractionpcsx2) [reply pcsx2/GS/Renderers/HW/GSTextureCache.cpp:396]: I didn't want the line to be too long, but sure, I guess.

**Review by kamfretoz — APPROVED** [review]

**Review by Mrlinkwii — APPROVED** [review]

**Comment by refractionpcsx2** [comment]
Want to try some better things instead of merging this.

**Comment by refractionpcsx2** [comment]
Superseeded (thankfully) by #10458

**Pull request closed without merging**